capnp/src/text.rs

```diff
@@ -39,12 +40,42 @@ impl crate::introspect::Introspect for Owned {
     }
 }
 
-pub type Reader<'a> = &'a str;
+#[derive(Clone, Copy, PartialEq)]
+pub struct Reader<'a> {
+    pub reader: data::Reader<'a>,
+}
+
+pub fn new_reader(v: &str) -> Reader<'_> {
+    Reader {
+        reader: v.as_bytes(),
+    }
+}
+
+impl<'a> Reader<'a> {
+    pub fn as_bytes(self) -> &'a [u8] {
+        self.reader
+    }
+
+    pub fn to_str(self) -> Result<&'a str> {
```

Comment:
https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

Looks to me like `as_str()` is probably the idiomatic thing.

Reply:
Oh, wait... `to_str()` looks right after all, because the utf-8 check means that the operation is not 'free':

> [Path::to_str](https://doc.rust-lang.org/std/path/struct.Path.html#method.to_str) performs an expensive UTF-8 check on the bytes of an operating system path. The input and output are both borrowed. It would not be correct to call this as_str because this method has nontrivial cost at runtime.

Reply:
Yes I mostly used `as_bytes` to reflect that this one is lightweight, just sending back a reference to the underlying bytes, whereas the `to_str` is slightly more involved, as it can fail so it shouldn't just be treated as a view of the reader's data. Happy to go to a consistent naming though, if you have a preference between one of the 2.

Reply:
My preference is to keep it as you have done here: `as_bytes()` and `to_str()`. It just took me a while to catch up on the justification for that. :)